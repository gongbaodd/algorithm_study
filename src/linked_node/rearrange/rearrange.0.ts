import { LNode } from "../lib";

export function reverse(_node: LNode): LNode {
  const node = _node;

  if (node.next === null) {
    return node;
  }

  const newFirst = reverse(node.next);
  node.next.next = node;
  node.next = null;

  return newFirst;
}

export function rearrage(head: LNode): LNode {
  if (!head.next) {
    return head;
  }

  let slowCursor = head.next;
  let fastCursor = head.next.next;

  while (fastCursor) {
    slowCursor = slowCursor.next as LNode;
    fastCursor = fastCursor.next?.next ?? null;
  }

  type nextNode = PropType<LNode, "next">;

  let tailCursor: nextNode = reverse(slowCursor);
  let headCursor: nextNode = head.next;

  for (;;) {
    const headNext: nextNode = headCursor?.next ?? null;
    const tailNext: nextNode = tailCursor?.next ?? null;

    const leftCursors = [headCursor, tailCursor].filter(
      cursor => cursor !== null
    );

    if (leftCursors.length === 0) {
      break;
    }

    if (leftCursors.length === 1) {
      (leftCursors[0] as LNode).next = null;
      break;
    }

    (headCursor as LNode).next = tailCursor;
    (tailCursor as LNode).next = headNext;

    headCursor = headNext;
    tailCursor = tailNext;
  }

  return head;
}
