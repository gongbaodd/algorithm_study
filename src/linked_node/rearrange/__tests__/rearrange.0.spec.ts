import { init } from "../../lib";
import { rearrage } from "../rearrange.0";

describe("Rearrange", () => {
  describe("empty chain", () => {
    it("should return empty chain", () => {
      const head = init([]);
      const newHead = rearrage(head);
      expect(newHead).toMatchSnapshot();
    });
  });

  describe("The length of the chain is odd", () => {
    it("should return as expected", () => {
      const head = init([1, 2, 3, 4, 5, 6, 7]);
      const newHead = rearrage(head);
      expect(newHead).toMatchSnapshot();
    });
  });

  describe("The length of the chain is even", () => {
    it("should return as expected", () => {
      const head = init([1, 2, 3, 4, 5, 6, 7, 8]);
      const newHead = rearrage(head);
      expect(newHead).toMatchSnapshot();
    });
  });
});
