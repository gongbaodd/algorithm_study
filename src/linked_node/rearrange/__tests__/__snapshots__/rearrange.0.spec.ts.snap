// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rearrange The length of the chain is even should return as expected 1`] = `
LNode {
  "data": 0,
  "next": LNode {
    "data": 1,
    "next": LNode {
      "data": 8,
      "next": LNode {
        "data": 2,
        "next": LNode {
          "data": 7,
          "next": LNode {
            "data": 3,
            "next": LNode {
              "data": 6,
              "next": LNode {
                "data": 4,
                "next": LNode {
                  "data": 5,
                  "next": null,
                },
              },
            },
          },
        },
      },
    },
  },
}
`;

exports[`Rearrange The length of the chain is odd should return as expected 1`] = `
LNode {
  "data": 0,
  "next": LNode {
    "data": 1,
    "next": LNode {
      "data": 7,
      "next": LNode {
        "data": 2,
        "next": LNode {
          "data": 6,
          "next": LNode {
            "data": 3,
            "next": LNode {
              "data": 5,
              "next": LNode {
                "data": 4,
                "next": null,
              },
            },
          },
        },
      },
    },
  },
}
`;

exports[`Rearrange empty chain should return empty chain 1`] = `
LNode {
  "data": 0,
  "next": null,
}
`;
